File: algo-c-to-_/src/multiply.py
#
#	from src/multiply.c
#
#	unsigned multiply(unsigned, unsigned)	to	mulA, mulB, mulC
#

def mulA(a, b):
    r = 0
    while a != 0:
        if a % 2 == 1:
            r += b
        b, a = b * 2, a // 2
    return r

def mulB(a, b):
    r = 0
    while a != 0:
        if a & 1 == 1:
            r += b
        b, a = b << 1, a >> 1
    return r

def mulC(a, b):
    def _iter(a, b, r):
        return _iter(a>>1, b<<1, r+b if a&1==1 else r) if a != 0 else r
    return _iter(a, b, 0)

File: porting-from-algo-c/examples/multiply.lua
--
--	from src/multiply.c
--
--	unsigned multiply(unsigned, unsigned)	to	iMulA, iMulB, iMulC
--

local M = require 'multiply'

local iMulA, iMulB, iMulC = M.iMulA, M.iMulB, M.iMulC

do
	print(("%s, %s, %s, %s -> %d, %d, %.2f, %.2f"):format(
		"2*3", "iMulA(2,3)", "2.1*3.1", "iMulA(2.1,3.1)", 2*3, iMulA(2,3), 2.1*3.1, iMulA(2.1,3.1)))
	print(("%s, %s, %s, %s -> %d, %d, %.2f, (_CAN'T_CALL_)"):format(
		"2*3", "iMulB(2,3)", "2.1*3.1", "iMulB(2.1,3.1)", 2*3, iMulB(2,3), 2.1*3.1))
	print(("%s, %s, %s, %s -> %d, %d, %.2f, (_CAN'T_CALL_)"):format(
		"2*3", "iMulC(2,3)", "2.1*3.1", "iMulC(2.1,3.1)", 2*3, iMulC(2,3), 2.1*3.1))

	--
	-- print(("%s, %s, %s, %s -> %d, %d, %.2f, %.2f"):format(
	-- 	"2*3", "iMulB(2,3)", "2.1*3.1", "iMulB(2.1,3.1)", 2*3, iMulB(2,3), 2.1*3.1, iMulB(2.1,3.1)))
	--
	-- ...
	-- lua: src/multiply.lua:19: number (local 'a') has no integer representation
	-- stack traceback:
	--         src/multiply.lua:19: in function 'multiply.iMulB'
	--         examples/multiply.lua:15: in main chunk
	--         [C]: in ?
	--
end
